### DISCUSSION TITLE
why downstream in a list do not work ?

### DISCUSSION THREAD

**Discussion by raphaelauv** [apache/airflow#36088] [Q&A] [answered]
```python
    a = EmptyOperator(task_id='a')
    b = EmptyOperator(task_id='b')
    c = EmptyOperator(task_id='c')
    d = EmptyOperator(task_id='d')

    a >> [ c << b, d]

```
give
![image](https://github.com/apache/airflow/assets/10202690/cd8a0280-66a0-4cbe-b95f-391d2bd0efbc)

but 
```python
    a = EmptyOperator(task_id='a')
    b = EmptyOperator(task_id='b')
    c = EmptyOperator(task_id='c')
    d = EmptyOperator(task_id='d')

    a >> [ b >> c, d]
```

give

![image](https://github.com/apache/airflow/assets/10202690/6f87dbc2-e562-4500-a677-dd9c24054c7b)

I don't understand why it's different for airflow

thanks all

**Comment by Taragolis** [comment, marked as answer]
Bitshift overloading functions return task from he right part, e.g  `c << b` will return `b` and `b >> c` will return `c`

https://github.com/apache/airflow/blob/87b08ad0840a11d8cd5c0b5043d3a341b1a8f258/airflow/models/taskmixin.py#L97-L105

That mean your first example `a >> [ c << b, d]` "transformed" into the `a >> [b, d]`, and `a >> [ b >> c, d]` "transformed" into the `a >> [c, d]`

**Reply by raphaelauv** [reply]
thanks @Taragolis 

so we do not have any sugar to create the first dag only with `>>` in one liner

only option is that ?
```python
a >> [ b , d]
b >> c
```

**Reply by Taragolis** [reply]
I'm not sure that some shorter way exists.

**Reply by potiuk** [reply]
```
a >> [ b , d]; b >> c
```

Python supports semicolons as separators.

**Reply by potiuk** [reply]
(and you have one-liner)

**Reply by raphaelauv** [reply]
Thanks @potiuk 👍

It feel not-friendly that the first code is not giving the same dag than the second code

It would be a breaking change , but would be great to be able to express "complex" composition without multiple lines

**Reply by Taragolis** [reply]
> Python supports semicolons as separators.

`black` / `ruff` would be not happy about reformat one liner to the regular one

**Reply by Taragolis** [reply]
> It feel not-friendly that the first code is not giving the same dag than the second code

I'm not sure that there is any solution exists for that, mainly behaviour about return right part based on:
- Historical reason
- Make it possible to work with more common cases such as: `a >> b >> c >> etc` and `etc << c << b << a`

**Reply by raphaelauv** [reply]
the unit test that explicit the problem -> 

https://github.com/apache/airflow/pull/36107

**Reply by Taragolis** [reply]
Maybe it is more about something miss explanation into the documentation?
The result the different because of the `(t1 >> t2) != (t2 << t1)`, note this about return values of the methods.

**Reply by raphaelauv** [reply]
yeah , and from a user perspective I see that as a bug

same as if python would say `3 < 4 != 4 > 3`

**Reply by Taragolis** [reply]
Depend on perspective: `(1 / 2) != (2 / 1)`

**Reply by potiuk** [reply]
Well. We are abusing Python operator to do something it was not really intended to do, so I'd hardly say it's a bug. It's a property of it :)

**Reply by potiuk** [reply]
And @raphaelauv -> if you know how to correct it wihout violating basic Python assumptions - I think we are all ears. While I might agree it is not perfect, you can only say it's fixable if you have viable alternative. Maybe a complete change in those "shorthand" notations we use, maybe using different operators, maybe some other magic to make it happen.

all ears :)

But if there is no easy way to fix it, I'd say "it's a property" and treat is as a "good enough" case.

World  we live in is not perfect, And I think many of us stirives to make it better with every commit :) . But sometimes "good enough" is the only feasible thing we can accept. Just sometimes :D

**Reply by raphaelauv** [reply]
I have a fix but before trying to merge it , I want to include many tests on the current behavior to know explicitly what will change and if we decide to merge it and accept the breaking change 

could you please review these PR please

https://github.com/apache/airflow/pull/37161
https://github.com/apache/airflow/pull/37162
thanks
